### DISCUSSION TITLE
More accurate timing for rendering?

### DISCUSSION THREAD

**Discussion by audiomax-ai** [FluidSynth/fluidsynth#1260] [Q&A]
Hi all, 

I'm seeking more information about rendering midi files to audio with more precision on timing. I'm rendering files in Python, but when I import the audio loops into a music program like Ableton or Pro Tools, the length of the loop is slightly off.

Is it possible to achieve the same rendering precision like professional music softwares? Does the audio driver affect the accuracy?

Any insights or comments would be appreciated. Thanks!

Max

**Comment by derselbst** [comment]
Fluidsynth processes events with an accuracy of 64 samples per samplerate. Given a default samplerate of 44100 Hz this yields a precision of 64/44100 = 0.0015 s

If this is not accurate enough for your use-case, then sry, you're out of luck. I'm using MIDI loops in my own player, and they sound fine to me. 

Otherwise, providing a complete minimal reproducible example would be helpful to understand your problem.

**Comment by audiomax-ai** [comment]
If I render 8 bars at 120bpm, I get a length of 15.981133786848073 seconds when analyzed with librosa.  I'm trying to bulk process midi files to wav with a soundfont. 

This is my python script running on colab. ChatGPT made it.
```
import os
import subprocess

directory = 'wav_output'
files = os.listdir(directory)
for file in files:
    if file.endswith('.wav'):
        file_path = os.path.join(directory, file)
        os.remove(file_path)

sf2_file = 'soundfonts/gm-kit.sf2'
midi_dir = 'midi_processed'
output_dir = 'wav_output'

os.makedirs(output_dir, exist_ok=True)

def convert_to_wav(midi_file, output_file, gain):
    command = ['fluidsynth', '-ni', sf2_file, midi_file, '-F', output_file, '-r', '44100', '-o', 'audio.file.format=s24', '-g', '1', '-o', 'audio.driver=coreaudio', '-o', 'audio.period-size=64']
    subprocess.call(command)

gain = 1.0

for file_name in os.listdir(midi_dir):
    if file_name.endswith('.mid'):
        midi_file = os.path.join(midi_dir, file_name)
        output_file = os.path.join(output_dir, os.path.splitext(file_name)[0] + '.wav')
        convert_to_wav(midi_file, output_file, gain)
```

**Reply by jimhen3ry** [reply]
If I understand derselbst's post correctly, and it is very possible I do not, a nominal 1.0000 seconds of output might be as short as 0.9985 seconds. Therefore, 8 four beat bars at 120 bpm, 16.0000 nominal seconds, could be as short as 15.9760 seconds. The measured 15.9811 seconds appears to be within the accuracy that FluidSynth can deliver.

**Reply by derselbst** [reply]
Using the sample rate of 44100 Hz as found in the script, you are correct. And since the events are delivered via sample-timer, I think this error could accumulate all the way up to 15.9760 s in worst case. I would not expect this to happen though. These timing errors should compensate each other more or less, like in a Gaussian distribution. I think one should rather verify that the timing of these "bar events" is on the dot, i.e. comes at exactly the expected delta time in the MIDI file. The next step would be to increase the sample rate to e.g. 96kHz and see if it's getting more accurate. If it doesn't, we might have an issue.
